File: Assets/Bundles/Shared/Resources/TS/Util/CanvasAPI.lua
-- Compiled with unity-ts v2.1.0-75
-- * Pointer button.
local PointerButton
do
	local _inverse = {}
	PointerButton = setmetatable({}, {
		__index = _inverse,
	})
	PointerButton.LEFT = 0
	_inverse[0] = "LEFT"
	PointerButton.RIGHT = 1
	_inverse[1] = "RIGHT"
	PointerButton.MIDDLE = 2
	_inverse[2] = "MIDDLE"
end
-- * Pointer direction.
local PointerDirection
do
	local _inverse = {}
	PointerDirection = setmetatable({}, {
		__index = _inverse,
	})
	PointerDirection.DOWN = 0
	_inverse[0] = "DOWN"
	PointerDirection.UP = 1
	_inverse[1] = "UP"
end
-- * Hover state.
local HoverState
do
	local _inverse = {}
	HoverState = setmetatable({}, {
		__index = _inverse,
	})
	HoverState.ENTER = 0
	_inverse[0] = "ENTER"
	HoverState.EXIT = 1
	_inverse[1] = "EXIT"
end
local CanvasAPI
do
	CanvasAPI = setmetatable({}, {
		__tostring = function()
			return "CanvasAPI"
		end,
	})
	CanvasAPI.__index = CanvasAPI
	function CanvasAPI.new(...)
		local self = setmetatable({}, CanvasAPI)
		return self:constructor(...) or self
	end
	function CanvasAPI:constructor()
	end
	function CanvasAPI:Init()
		self.canvasUIEvents = GameObject:Find("CanvasUIEvents"):GetComponent("CanvasUIEvents")
		self.canvasHitDetector = self.canvasUIEvents.gameObject:GetComponent("CanvasHitDetector")
	end
	function CanvasAPI:RegisterEvents(gameObject)
		self.canvasUIEvents:RegisterEvents(gameObject)
	end
	function CanvasAPI:IsPointerOverUI()
		return self.canvasHitDetector:IsPointerOverUI()
	end
	function CanvasAPI:OnPointerEvent(targetGameObject, callback)
		self:Setup(targetGameObject)
		self.eventInterceptor:OnPointerEvent(function(instanceId, direction, button)
			-- Only run callback if instance ids match.
			if instanceId == targetGameObject:GetInstanceID() then
				callback(direction, button)
			end
		end)
	end
	function CanvasAPI:OnHoverEvent(targetGameObject, callback)
		self:Setup(targetGameObject)
		self.eventInterceptor:OnHoverEvent(function(instanceId, hoverState)
			-- Only run callback if instance ids match.
			if instanceId == targetGameObject:GetInstanceID() then
				callback(hoverState)
			end
		end)
	end
	function CanvasAPI:OnSubmitEvent(targetGameObject, callback)
		self:Setup(targetGameObject)
		self.eventInterceptor:OnSubmitEvent(function(instanceId)
			-- Only run callback if instance ids match.
			if instanceId == targetGameObject:GetInstanceID() then
				callback()
			end
		end)
	end
	function CanvasAPI:OnSelectEvent(targetGameObject, callback)
		self:Setup(targetGameObject)
		self.eventInterceptor:OnSelectEvent(function(instanceId)
			-- Only run callback if instance ids match.
			if instanceId == targetGameObject:GetInstanceID() then
				callback()
			end
		end)
	end
	function CanvasAPI:OnDeselectEvent(targetGameObject, callback)
		self:Setup(targetGameObject)
		self.eventInterceptor:OnDeselectEvent(function(instanceId)
			-- Only run callback if instance ids match.
			if instanceId == targetGameObject:GetInstanceID() then
				callback()
			end
		end)
	end
	function CanvasAPI:OnClickEvent(targetGameObject, callback)
		self:Setup(targetGameObject)
		self.eventInterceptor:OnClickEvent(function(instanceId)
			-- Only run callback if instance ids match.
			if instanceId == targetGameObject:GetInstanceID() then
				callback()
			end
		end)
	end
	function CanvasAPI:OnValueChangeEvent(targetGameObject, callback)
		self:Setup(targetGameObject)
		self.eventInterceptor:OnValueChangeEvent(function(instanceId, value)
			if instanceId == targetGameObject:GetInstanceID() then
				callback(value)
			end
		end)
	end
	function CanvasAPI:Setup(gameObject)
		if CanvasAPI.eventInterceptor == nil then
			self.eventInterceptor = GameObject:Find("CanvasUIEventsInterceptor"):GetComponent("CanvasUIEventInterceptor")
		end
		self:RegisterEvents(gameObject)
	end
end
return {
	PointerButton = PointerButton,
	PointerDirection = PointerDirection,
	HoverState = HoverState,
	CanvasAPI = CanvasAPI,
}
-- ----------------------------------
-- ----------------------------------

File: Assets/Bundles/Imports/Core/Core~/src/Shared/Util/CanvasAPI.ts
/** Pointer button. */
export enum PointerButton {
	LEFT,
	RIGHT,
	MIDDLE,
}

/** Pointer direction. */
export enum PointerDirection {
	DOWN,
	UP,
}

/** Hover state. */
export enum HoverState {
	ENTER,
	EXIT,
}

export class CanvasAPI {
	/** Global event interceptor. */
	private static eventInterceptor: CanvasUIEventInterceptor | undefined;

	private static canvasUIEvents: CanvasUIEvents;
	private static canvasHitDetector: CanvasHitDetector;

	public static Init(): void {
		this.canvasUIEvents = GameObject.Find("CanvasUIEvents").GetComponent<CanvasUIEvents>();
		this.canvasHitDetector = this.canvasUIEvents.gameObject.GetComponent<CanvasHitDetector>();
	}

	public static RegisterEvents(gameObject: GameObject): void {
		this.canvasUIEvents.RegisterEvents(gameObject);
	}

	public static IsPointerOverUI() {
		return this.canvasHitDetector.IsPointerOverUI();
	}

	/**
	 * Subscribe to pointer events for a given target. `targetGameObject` MUST have an `EventTrigger` component
	 * to be eligible to receive input events. See the `ShopItem` prefab for an example.
	 *
	 * @param targetGameObject Target of pointer event.
	 * @param callback Callback to run when `targetGameObject` is the subject of a pointer event. Includes direction and button.
	 */
	public static OnPointerEvent(
		targetGameObject: GameObject,
		callback: (direction: PointerDirection, button: PointerButton) => void,
	): void {
		this.Setup(targetGameObject);
		this.eventInterceptor!.OnPointerEvent((instanceId, direction, button) => {
			/* Only run callback if instance ids match. */
			if (instanceId === targetGameObject.GetInstanceID()) {
				callback(direction as PointerDirection, button as PointerButton);
			}
		});
	}

	/**
	 * Subscribe to hover events for a given target. `targetGameObject` MUST have an `EventTrigger` component
	 * to be eligible to receive input events. See the `ShopItem` prefab for an example.
	 *
	 * @param targetGameObject Target of hover event.
	 * @param callback Callback to run when `targetGameObject` is the subject of a hover event. Includes hover state.
	 */
	public static OnHoverEvent(targetGameObject: GameObject, callback: (hoverState: HoverState) => void): void {
		this.Setup(targetGameObject);
		this.eventInterceptor!.OnHoverEvent((instanceId, hoverState) => {
			/* Only run callback if instance ids match. */
			if (instanceId === targetGameObject.GetInstanceID()) {
				callback(hoverState as HoverState);
			}
		});
	}

	public static OnSubmitEvent(targetGameObject: GameObject, callback: () => void): void {
		this.Setup(targetGameObject);
		this.eventInterceptor!.OnSubmitEvent((instanceId) => {
			/* Only run callback if instance ids match. */
			if (instanceId === targetGameObject.GetInstanceID()) {
				callback();
			}
		});
	}

	public static OnInputFieldSubmit(targetGameObject: GameObject, callback: (data: string) => void): void {
		this.Setup(targetGameObject);
		this.eventInterceptor!.OnInputFieldSubmitEvent((instanceId, data) => {
			/* Only run callback if instance ids match. */
			if (instanceId === targetGameObject.GetInstanceID()) {
				callback(data);
			}
		});
	}

	public static OnSelectEvent(targetGameObject: GameObject, callback: () => void): void {
		this.Setup(targetGameObject);
		this.eventInterceptor!.OnSelectEvent((instanceId) => {
			/* Only run callback if instance ids match. */
			if (instanceId === targetGameObject.GetInstanceID()) {
				callback();
			}
		});
	}

	public static OnDeselectEvent(targetGameObject: GameObject, callback: () => void): void {
		this.Setup(targetGameObject);
		this.eventInterceptor!.OnDeselectEvent((instanceId) => {
			/* Only run callback if instance ids match. */
			if (instanceId === targetGameObject.GetInstanceID()) {
				callback();
			}
		});
	}

	public static OnClickEvent(targetGameObject: GameObject, callback: () => void): void {
		this.Setup(targetGameObject);
		this.eventInterceptor!.OnClickEvent((instanceId) => {
			/* Only run callback if instance ids match. */
			if (instanceId === targetGameObject.GetInstanceID()) {
				callback();
			}
		});
	}

	public static OnValueChangeEvent(targetGameObject: GameObject, callback: (value: number) => void): void {
		this.Setup(targetGameObject);
		this.eventInterceptor!.OnValueChangeEvent((instanceId, value) => {
			if (instanceId === targetGameObject.GetInstanceID()) {
				callback(value);
			}
		});
	}

	/** Fetches and sets the global event interceptor. */
	private static Setup(gameObject: GameObject): void {
		if (CanvasAPI.eventInterceptor === undefined) {
			this.eventInterceptor =
				GameObject.Find("CanvasUIEventsInterceptor").GetComponent<CanvasUIEventInterceptor>();
		}
		this.RegisterEvents(gameObject);
	}
}
